### PR TITLE
Add rustc guide to toolstate

### LINKED ISSUE #59597: Tracking Issue: Add rustc-guide to toolstate
(According to the plan in https://github.com/rust-lang-nursery/rust-toolstate/issues/5)

cc @nikomatsakis @kennytm @Michael-F-Bryan

@kennytm Could you tag this with Help-Wanted?

# Instructions

**Background**: We have a cool tool called `toolstate` which posts a message on PRs that break dependent crates (e.g. `clippy`). For example, it might generate a post [like this](https://github.com/rust-lang/rust/pull/59298#issuecomment-474733101). There is also a web portal: https://rust-lang-nursery.github.io/rust-toolstate/

**Objective**: Currently the [rustc-guide](https://github.com/rust-lang/rustc-guide) is not tracked by toolstate, but we want it to be. This will help us detect when the compiler changes and the guide needs to be updated. However, some infrastructure work is needed for this to work.

**Steps**
- [ ] Add `mdbook` and `mdbook-linkcheck` as dependencies for the rust-lang/rust build. These are used to build and check links in the `rustc-guide` repo.
    - TBH, I don't know how best to do this. We already use `mdbook` to build The Book and other things, but it's not clear to me how we make sure they are installed... (Perhaps @kennytm has ideas?)
- [ ] Add an `./x.py test src/doc/rustc-guide` command
      1. Create a new `struct RustcGuide` in `src/bootstrap/test.rs`.
      2. `impl Step for RustcGuide` to run `mdbook build` in `src/doc/rustc-guide`
           - Make sure that this test is not part of the standard build process or it will block PRs on the rust-lang/rust repo. IIUC, this can be done in the implementation of `should_run` and `make_run` in `Step` (see the documentation in `src/bootstrap/builder.rs`
      3. Update `Builder::get_step_descriptions`
      4. You may or may not need to do other stuff to wire things together... not sure. In the end, you should be able to do `./x.py test src/doc/rustc-guide` to run the book build and linkchecker on the guide. 
- [ ] Update the toolstate config to add `rustc-guide`, as done here: https://github.com/rust-lang/rust/pull/59038/files
    - Add me (@mark-i-m) under the `MAINTAINERS` section. (@nikomatsakis do you want to be added too?) 

Feel free to ping me on this thread if you need help :)

### PR DESCRIPTION
**Pull request by andrehjr**
Closes #59597

### CODE REVIEW

**Comment by mark-i-m** [comment]
Hmm it looks like `libssl-dev` is not installed on the Linux build machines. You might need to edit the .travis.yml file to make sure it is installed as a dependency of the build.

**Comment by andrehjr** [comment]
I'll take a look into that!

**Comment by mark-i-m** [comment]
Hmm... you might need to install both `libssl` and `libssl-dev`?

**Comment by andrehjr** [comment]
It seems to me, that libssl-dev and stuff are already installed on the linux machine.. but somehow pkg-config is not being able to pick it up on that part of the build process.. I could try something like `you can set the `OPENSSL_DIR` environment variable for the compilation process.` but it seems too much 🤔

**Comment by mark-i-m** [comment]
@sfackler Do you know what might be causing this?

**Comment by sfackler** [comment]
Doesn't the build run in the docker image? Installing things on the Travis host isn't going to work in that case.

**Comment by andrehjr** [comment]
That's true, the build runs into the docker image. I added the required deps onto the image and the build finally finished with success now! Thanks! ✨ 

@mark-i-m what are the next steps?

File: src/tools/rustbook/src/main.rs
Comment (mark-i-m) [inline src/tools/rustbook/src/main.rs]: Could we rename `broken_links` to `err`? It is not guaranteed to be a broken links error; there are a bunch of other error types: https://docs.rs/mdbook-linkcheck/0.2.3/mdbook_linkcheck/errors/index.html.

File: src/tools/rustbook/src/main.rs
Comment (mark-i-m) [inline src/tools/rustbook/src/main.rs]: This will crash if the error is not a `BrokenLinks`. I would propose using `downcast` instead of `downcast_ref`. In the case where it is a `BrokenLinks` we can just print the links (as you already do). Otherwise, we can just print the error as you do on line 77.

**Review by mark-i-m — CHANGES_REQUESTED** [review]
@andrehjr Nice work! I left a few comments below, but overall it looks good to me.

The next steps would be to un-WIP the title and wait for a review from somebody with bors permissions (I only have permissions on the rustc-guide repo). This should happen relatively soon. At that point, bors will test and merge PR, and that's it!

**Review by mark-i-m — APPROVED** [review]
Sorry, I missed one last thing. Other than that I think this is ready 🎉

**Comment by mark-i-m** [comment]
My pleasure :) Thanks for contributing.

**Comment by mark-i-m** [comment]
@andrehjr perhaps you could post `r? @kennytm`? They  might be to get to your PR faster?

**Comment by andrehjr** [comment]
r? @kennytm

Sure!

**Comment by mark-i-m** [comment]
@andrehjr It looks like this needs a rebase.

@kennytm Ping?

**Review by kennytm — APPROVED** [review]
r=me after fixing Cargo.lock

**Comment by andrehjr** [comment]
@bors r=kennytm

Cargo.lock is fixed!

**Comment by andrehjr** [comment]
r=kennytm

**Comment by mark-i-m** [comment]
@andrehjr Only people on bors's list can issue commands to bors (or I would have done so a while back :P).

**Comment by andrehjr** [comment]
@mark-i-m yes lol I thought it would work after a r=me, I'm new to bors :) 

Thanks @mark-i-m @kennytm !

**Comment by andrehjr** [comment]
Ok, so, judging by the error on tidy. The issue happens because  [uuid's Cargo.toml](https://github.com/uuid-rs/uuid/blob/master/Cargo.toml#L28), which lists the license as 'Apache-2.0 OR MIT`.

And only 'MIT OR Apache-2.0' is whitelisted on the [src/tools/tidy/src/deps.rs](https://github.com/rust-lang/rust/blob/master/src/tools/tidy/src/deps.rs#L11) lol..

uuid is pulled as a dep of reqwest/mdbook-linkcheck.

So, I see the following possible solutions:
1) Add the `Apache-2.0 OR MIT` variation to the LICENSES on tidy/src/deps.rs. But there might be more libs with the same problem. Hopefully not.
2) Add mdbook-linkcheck and its deps to EXCEPTIONS on tidy/src/deps.rs -- which does not feel right, looking at the comment [here](https://github.com/rust-lang/rust/blob/master/src/tools/tidy/src/deps.rs#L22-L25), but as these are deps of rust tooling, so, maybe it's alright?

Thoughts @mark-i-m ?

**Comment by mark-i-m** [comment]
Thanks @andrehjr for investigating :)

(1) seems like the right option for this PR, but in the long run, matching the license string seems like a flawed design...

**Comment by mark-i-m** [comment]
Ping @kennytm

**Comment by kennytm** [comment]
🤷‍♀️ Please rebase again. 

@bors p=1

**Comment by andrehjr** [comment]
Rebased again @kennytm !

**Comment by andrehjr** [comment]
I'll need to backtrack a bit and ensure libssl-dev deps exist on all images to make this work involved during the tests :( Seems like some images already have it, others don't. Sorry about that! 

The annoying part is that this is only getting caught on the 'bigger' build.

Is making that change acceptable? Feels like the scope of adding this to the toolstate is getting bigger and bigger  @kennytm @mark-i-m :/

**Comment by mark-i-m** [comment]
@andrehjr TBH, I don't know. I had not expected it to be this involved. Maybe it makes sense to do it in another PR and block this PR on that one? @kennytm do you have any thoughts?

**Comment by kennytm** [comment]
Could we use rustls instead of OpenSSL? Or could we just run the doc-test in this PR, and add the linkcheck in another PR?

**Comment by mark-i-m** [comment]
I think OpenSSL is a transitive dependency, so we don't have much control. Also, the linkchecker was the main objective of adding rustc-guide to toolstate. We don't actually have many doctests in the guide.

**Comment by mark-i-m** [comment]
@kennytm What do you think is the path forward? Should I try to open a PR to add openssl?

**Comment by kennytm** [comment]
Let's do it in this PR. It would  be strange to have a separate PR that adds OpenSSL dependency to every image without the rationale.

**Comment by andrehjr** [comment]
I'll try doing that this weekend ✨ @mark-i-m

**Comment by andrehjr** [comment]
@mark-i-m I added libssl-dev/pkg-config to all images that needed! I think it should be all good now

A lot of them already had it via a shared script on `cross-apt-packages.sh` or installing directly just like I did.

**Comment by mark-i-m** [comment]
@andrehjr Thanks :+1: 

@kennytm ready for review :smiley_cat:

**Review by mark-i-m — APPROVED** [review]

**Comment by mark-i-m** [comment]
Ping @kennytm This is still waiting for your review. It would be immensely useful to get this merged.

**Comment by andrehjr** [comment]
Rebased again! ✨

**Comment by mark-i-m** [comment]
> /rustroot/bin/ld: skipping incompatible /rustroot/lib/libssl.so when searching for -lssl

???

**Comment by kennytm** [comment]
@bors r-

Perhaps those libraries are not 32-bit?

Or maybe it's less a hassle to let `mdbook-linkcheck` expose the features of `reqwest` so we could use `rustls-tls` 🤷 (cc @Michael-F-Bryan)

**Comment by mark-i-m** [comment]
Is there a way to just run this on the x86-64-linux build? This is target independent and we don't really need it running on every container.

**Comment by kennytm** [comment]
Maybe you could add a `#[cfg(feature = ...)]` to the rustbook tool, and turn the link checker into an optional dependency.

**Comment by mark-i-m** [comment]
The problem is that there are other users of the rustbook tool, and they might not be aware of this strange special-casing behavior... I will try to look into this a bit more later this week.

**Comment by mark-i-m** [comment]
A quick heads up: https://github.com/rust-lang/rust/pull/61948 will break this PR because linkcheck is not compatible with mdbook 0.3. I have opened https://github.com/Michael-F-Bryan/mdbook-linkcheck/pull/9 to fix this...

**Comment by Michael-F-Bryan** [comment]
@kennytm and @mark-i-m this should be unblocked now. I've released mdbook-linkcheck v0.3.0.

**Comment by mark-i-m** [comment]
@andrehjr Thanks for your patience. I finally got a chance to look into this. I think @kennytm's suggestion of using a `cfg` is the most viable path forward. I opened a PR on your branch: https://github.com/andrehjr/rust/pull/1. Let me know what you think. Also, there are conflicts with rust-lang/rust#master that need to be resolved.

**Comment by mark-i-m** [comment]
Lol already a conflict...

**Comment by mark-i-m** [comment]
ping @kennytm perhaps it will work now?

**Comment by mark-i-m** [comment]
I never expected this PR to be so much effort to get it to merge... @andrehjr @kennytm Thanks for your perseverance.

**Comment by kennytm** [comment]
@bors r-

Failed to build on macOS (perhaps everywhere not x86_64 + Linux

**Comment by mark-i-m** [comment]
@andrehjr I opened https://github.com/andrehjr/rust/pull/2 to fix the build errors, but I don't have a Mac to test on, so I'm not sure if there will be other errors after that.

**Comment by andrehjr** [comment]
Cool, I tested on macos @mark-i-m ! Looks like it's a success now!

```
   Compiling rustbook v0.1.0 (/Users/andrehjr/work/rust/src/tools/rustbook)
    Finished release [optimized] target(s) in 3.59s
Error: mdbook-linkcheck only works on x86_64 linux targets.

command did not execute successfully: "/Users/andrehjr/work/rust/build/x86_64-apple-darwin/stage0-tools-bin/rustbook" "linkcheck" "/Users/andrehjr/work/rust/src/doc/rustc-guide"
expected success, got: exit code: 101


failed to run: /Users/andrehjr/work/rust/build/bootstrap/debug/bootstrap test src/doc/rustc-guide
Build completed unsuccessfully in 0:00:05
```

**Comment by mark-i-m** [comment]
Hmm.... actually, I think it was a mistake to return an error. That will cause it to fail the build. @andrehjr Could you please change the `bail` to a println?

**Comment by andrehjr** [comment]
That makes sense @mark-i-m ! I pushed a fix 🚀

**Comment by mark-i-m** [comment]
@Centril I understand the importance of the rollups, but could you avoid preempting running builds of PRs with 100s of comments please?

@bors r+

**Comment by Centril** [comment]
@mark-i-m 

> @Centril I understand the importance of the rollups, but could you avoid preempting running builds of PRs [...]?

Sorry, but no. Maintaining high PR throughput and low latency (a shorter queue) is the 1. priority when managing the bors queue, especially after last week's CI problems after which we had a huge number of PRs in the queue. When one of my rollups fail due to some PR in it or some spurious failure I won't let other PRs leap ahead while I remake the rollup unless there is some special circumstance.

> [...] with 100s of comments please?

What's the relevance of 100s of comments?

**Comment by mark-i-m** [comment]
@Centril the significance is that a disproportionately high amount of work and waiting has gone into this PR.

Moreover, it seems like every time the PR is approved, it incurs a merge conflict; it's been rebased at least 5 times in the last two weeks. It's just a bit frustrating when such a PR finally makes it to the head of the queue... and is preempted...

**Comment by Centril** [comment]
> @Centril the significance is that a disproportionately high amount of work and waiting has gone into this PR.

The queue system works such that older PRs automatically get ahead in the queue. The PR has also tested 4 times and failed. To me this means that this PR has been treated fairly by the queue.

> Moreover, it seems like every time the PR is approved, it incurs a merge conflict; it's been rebased at least 5 times in the last two weeks.

When the queue size is large it contributes towards having to rebase more times but if some PR is especially bitrotty you should raise the priority and write a comment saying it is bitrotty. In this case rebases were necessary due to lockfiles which are updated primarily outside of rollups by PRs that jump the queue with `p=1`.

> It's just a bit frustrating when such a PR finally makes it to the head of the queue... and is preempted...

First, a PR has a better chance of being tested if the queue is small and this is achieved through rollups. Second, it is equally frustrating to baby sit rollups, have them fail, and then be preempted by a PR that just got added to the queue. The alternative would be to treeclose while doing rollup ops but that doesn't seem to work so well.

**Pull request merged**
